streams/src/main/java/org/apache/kafka/streams/processor/internals/Tasks.java

```diff
@@ -103,42 +101,75 @@ public void addPendingStandbyTasksToCreate(final Map<TaskId, Set<TopicPartition>
+            return pendingUpdateActions.remove(taskId).getInputPartitions();
+        }
+        return null;
     }
 
     @Override
     public void addPendingTaskToRecycle(final TaskId taskId, final Set<TopicPartition> inputPartitions) {
-        pendingTasksToRecycle.put(taskId, inputPartitions);
+        pendingUpdateActions.put(taskId, PendingUpdateAction.createRecycleTask(inputPartitions));
     }
 
     @Override
     public Set<TopicPartition> removePendingTaskToUpdateInputPartitions(final TaskId taskId) {
-        return pendingTasksToUpdateInputPartitions.remove(taskId);
+        if (containsTaskIdWithAction(taskId, Action.UPDATE_INPUT_PARTITIONS)) {
+            return pendingUpdateActions.remove(taskId).getInputPartitions();
+        }
+        return null;
     }
 
     @Override
     public void addPendingTaskToUpdateInputPartitions(final TaskId taskId, final Set<TopicPartition> inputPartitions) {
-        pendingTasksToUpdateInputPartitions.put(taskId, inputPartitions);
+        pendingUpdateActions.put(taskId, PendingUpdateAction.createUpdateInputPartition(inputPartitions));
     }
 
     @Override
     public boolean removePendingTaskToCloseDirty(final TaskId taskId) {
-        return pendingTasksToCloseDirty.remove(taskId);
+        if (containsTaskIdWithAction(taskId, Action.CLOSE_DIRTY)) {
+            pendingUpdateActions.remove(taskId);
+            return true;
+        }
+        return false;
     }
 
     @Override
     public void addPendingTaskToCloseDirty(final TaskId taskId) {
-        pendingTasksToCloseDirty.add(taskId);
+        pendingUpdateActions.put(taskId, PendingUpdateAction.createCloseDirty());
     }
 
     @Override
     public boolean removePendingTaskToCloseClean(final TaskId taskId) {
-        return pendingTasksToCloseClean.remove(taskId);
+        if (containsTaskIdWithAction(taskId, Action.CLOSE_CLEAN)) {
+            pendingUpdateActions.remove(taskId);
+            return true;
+        }
+        return false;
     }
 
     @Override
     public void addPendingTaskToCloseClean(final TaskId taskId) {
-        pendingTasksToCloseClean.add(taskId);
+        pendingUpdateActions.put(taskId, PendingUpdateAction.createCloseClean());
+    }
+
+    @Override
+    public boolean removePendingActiveTaskToSuspend(final TaskId taskId) {
+        if (containsTaskIdWithAction(taskId, Action.SUSPEND)) {
+            pendingUpdateActions.remove(taskId);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void addPendingActiveTaskToSuspend(final TaskId taskId) {
+        pendingUpdateActions.put(taskId, PendingUpdateAction.createSuspend());
+    }
+
+    private boolean containsTaskIdWithAction(final TaskId taskId, final Action action) {
+        final PendingUpdateAction pendingUpdateAction = pendingUpdateActions.get(taskId);
+        return !(pendingUpdateAction == null || pendingUpdateAction.getAction() != action);
```

Comment:
nit: why not put the `!` inside as `pendingUpdateAction != null && pendingUpdateAction.getAction() == action`?

Reply:
Done in https://github.com/apache/kafka/pull/12638